### PR TITLE
irbempy: Improve test coverage

### LINKED ISSUE #307: test coverage: irbempy
irbempy init is 100% coverage, but the real work in irbempy is 45% covered. 

- find_LCDS is untested, as is find_LCDS_K. I wonder if we could use lgmpy to determine "truth" and compare.
- get_AEP8 has a lot of exceptional cases not handled, but prints out. Maybe these should be warnings or exceptions.
- _get_Lstar when nAlpha == 0
- get_Lm --again, can use lgmpy, as that's pretty heavily tested for this.
- get_Lstar

Hard to test:
- updateTS07Coeffs downloads data and could probably be hit with an integration test

### PR DESCRIPTION
**Pull request by jamesw732**
This pull request improves the test coverage of the irbempy submodule, up to 88% from about 80%. Also, some minor bug fixes and code improvements along the way. Closes #307 and (slightly off-topic) #768.

The bulk of the improvements came from testing cases where the internal irbemlib call returned close to badval, and the further handling as np.nan. I also added some regression tests for untested edge cases and in some cases, complete functions (`find_Lm`). All the new tests are regression tests since I do not have a first principles understanding of the code.

What's still missing is some harder-to-hit edge cases and the hard-to-test header code.

Unfortunately, the edge cases in `find_LCDS` and `find_LCDS_K` mean those functions have to be run several times, causing`test_irbempy.py` to run very slowly - about 5 minutes on Windows, 87 seconds on WSL (you read that right). Given talks about deprecating `find_LCDS` in #768, it might be safe to just remove or skip its tests for the sake of improving the runtime of the tests.

Originally, I intended to include a fix for #750 in this PR but decided that it is too off-topic and might require too much targeted review, so it makes sense to just put that in another PR that depends on this one.

## PR Checklist
- [X] Pull request has descriptive title
- [X] Pull request gives overview of changes
- [X] New code has inline comments where necessary
- [X] Any new modules, functions or classes have docstrings consistent with SpacePy style
- [X] (N/A) Added an entry to release notes if fixing a significant bug or providing a new feature
- [X] (N/A) New features and bug fixes should have unit tests
- [X] Relevant issues are linked to (e.g. `See issue #` or `Closes #`)

### CODE REVIEW

**Comment by jamesw732** [comment]
Since the CI still fails on main and this is otherwise ready for review, I figure it doesn't hurt to mark it as so.

**Comment by jtniehof** [comment]
Thanks James. Looks like the only problem is the known one fixed in #759. I hope to get a better look Monday, but you can also look at rebasing possibilities if there's some back-and-forth and blind alleys in history.

File: spacepy/irbempy/__init__.py [hunk 1]
```diff
@@ -894,7 +850,7 @@ def find_LCDS_K(ticks, K, extMag='T01STORM', options=[1, 1, 3, 0, 0], omnivals=N
                 continue
 
             # print('L* at inner bracket: {0}'.format(LS1['Lstar']))
-            LCDS, LCDS_PA = LS1['Lstar'][0], pa
+            LCDS, LCDS_PA = LS1['Lstar'][0, 0], pa[0]
```
Comment (jtniehof) [inline spacepy/irbempy/__init__.py:853]: This feels inconsistent because pa is treated as a scalar above, in the assignment to results. Does AlphaOfK return a single value or an array of length 1? Should probably document it.

File: tests/test_irbempy.py [hunk 2]
```diff
@@ -108,6 +108,20 @@ def test_prep_irbem(self):
             numpy.testing.assert_almost_equal(expected[key],
                                               actual[key],
                                               decimal=5)
+    def test_prep_irbem_sph_coords(self):
+        """Test prep_irbem with spherical coordinate input"""
+        loci = spacepy.coordinates.Coords([[3,0,0],[2,0,0]], 'GDZ', 'sph', use_irbem=True)
```
Comment (jtniehof) [inline tests/test_irbempy.py:113]: Can we try something with nontrivial angles here?

File: tests/test_irbempy.py [hunk 3]
```diff
@@ -314,6 +396,52 @@ def test_find_footpoint(self):
...
+    def test_get_Bfield_badval(self):
+        """test get_Bfield hitting badval"""
+        loci_noncredible = spacepy.coordinates.Coords([[-1e4, 1e4, 0], [2, 0, 0]], 'GEO', 'car', use_irbem=True)
+        actual = ib.get_Bfield(self.ticks, loci_noncredible, omnivals=self.omnivals)
+        self.assertTrue(np.isnan(actual["Blocal"][0]))
+        self.assertTrue(np.isnan(actual["Bvec"][0]).all())
+
+    def test_find_Bmirror_badval(self):
+        """test find_Bmirror hitting badval"""
+        loci_noncredible = spacepy.coordinates.Coords([[-1e4, 1e4, 0], [2, 0, 0]], 'GEO', 'car', use_irbem=True)
```
Comment (jtniehof) [inline tests/test_irbempy.py:412]: Can we try this where the location is credible, but the pitch angle is in the loss cone? Here (if I'm reading this right) the PA 0 case is also the "way out in the Universe" case.

File: tests/test_irbempy.py [hunk 4]
```diff
@@ -347,6 +475,19 @@ def test_coord_trans(self):
             [ 1.91462214,  0.06992421,  0.57387514]])
         numpy.testing.assert_almost_equal(expected, ib.coord_trans(self.loci, 'GSM', 'car'))
 
+    def test_coord_trans_swap(self):
+        """test coord_trans with invalid sysaxesout requiring a carsph swap"""
+        # re-assign ticks so this test can be run in isolation
+        self.loci.ticks = self.ticks
```
Comment (jtniehof) [inline tests/test_irbempy.py:481]: I'm not sure what this means. It's assigning ticks to the loci, which didn't have it before, but how is that isolating? (The setUp is run for every test, so one test shouldn't affect another)

**Review by jtniehof — COMMENTED** [review]
Few suggestions on a quick skim

File: spacepy/irbempy/__init__.py (hunk 1 above)
Comment (jamesw732) [reply spacepy/irbempy/__init__.py:853]: Aye it's always an array. This was causing a previously unhit numpy deprecation warning (and somehow an error in the CI... that was a frustrating one until I read more into the deprecation warning).

File: tests/test_irbempy.py (hunk 2 above)
Comment (jamesw732) [reply tests/test_irbempy.py:113]: Yes, I'm kind of surprised I didn't notice that was missing from the prep_irbem test.

File: tests/test_irbempy.py (hunk 3 above)
Comment (jamesw732) [reply tests/test_irbempy.py:412]: Yes, I think the reason I used a non-credible location was because I couldn't make Blocal hit badval otherwise. It seems appropriate to me to split this test up into two cases: noncredible location with a reasonable pitch angle, and credible location with zero pitch angle.

File: tests/test_irbempy.py (hunk 4 above)
Comment (jamesw732) [reply tests/test_irbempy.py:481]: Somehow, I was not aware of that last part. I'll remove that comment.

File: spacepy/irbempy/__init__.py (hunk 1 above)
Comment (jtniehof) [reply spacepy/irbempy/__init__.py:853]: But every place else pa is treated as a scalar: `results['AlphaEq'][idxt, idxk] = pa`. It's going to be `nan` if `GEOcoord[0]` isn't finite (which is a bit weird but I guess the assumption is that all three dimensions will be bad for a bad coordinate).

Do we treat this as "irbempy needs more work and this is out of scope for this PR"?

**Review by jtniehof — APPROVED** [review]
I'm going to merge this, but it leaves two open questions:
1. What further irbempy cleanup do we need to do, and do we need to make notes of all of them or just figure we'll find them back?
2. What should we put in release notes? find_Bmirror always returning the same locus sounds fairly severe but nobody's tripped over it yet...do we just say "several fixes for unusual cases in irbempy" or something? @drsteve ?
